### PR TITLE
Add authorisation to relayer denyRequest endpoint

### PR DESCRIPTION
**Pull request by gasiortomasz**

### CODE REVIEW

File: universal-login-sdk/lib/RelayerApi.ts [hunk 1]
```diff
@@ -40,10 +40,13 @@ export class RelayerApi {
   }
 
   async denyConnection(walletContractAddress: string, key: string) {
-    return this.http('POST', `/authorisation/${walletContractAddress}`, {
+    console.log('in relayerapi result');
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/RelayerApi.ts]: bleh

File: universal-login-sdk/lib/RelayerApi.ts [hunk 2]
```diff
@@ -40,10 +40,13 @@ export class RelayerApi {
...
+    const result = this.http('POST', `/authorisation/${walletContractAddress}`, {
       walletContractAddress,
       key,
     });
+    console.log('in relayerapi result', result);
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/RelayerApi.ts]: buuuu ;)

File: universal-login-commons/lib/core/models/authorisation.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export interface CancelAuthorisationRequest {
+  walletContractAddress: string;
+  key: string;
+}
+
+// TODO CancelAuthorisationRequest
+// TODO GetAuthorisationRequest
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/models/authorisation.ts]: no comments in master branch

File: universal-login-relayer/lib/core/utils/errors.ts [hunk 4]
```diff
@@ -77,6 +77,13 @@ export class InvalidMessage extends NotFound {
   }
 }
 
+export class InvalidAddress extends NotFound {
+  constructor (address: string) {
+    super(`Could not find address: ${address} in Multisig Wallet` , 'InvalidExecution');
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/core/utils/errors.ts]: `Unauthorised address: ${address}`

File: universal-login-relayer/lib/core/utils/errors.ts [hunk 5]
```diff
@@ -77,6 +77,13 @@ export class InvalidMessage extends NotFound {
   }
 }
 
+export class InvalidAddress extends NotFound {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/core/utils/errors.ts]: We should add new base exception `Unauthorised`, from which we should inherit, which should be handled with 401 http code in `errorHandler` here:
https://github.com/UniversalLogin/UniversalLoginSDK/blob/17c3cfd0291e965aed13692721ac8b11bcf39935/universal-login-relayer/lib/http/middlewares/errorHandler.ts#L34

File: universal-login-commons/lib/core/utils/authorisation.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import {utils} from 'ethers';
+
+import {CancelAuthorisationRequest} from '../models/authorisation';
+
+export const hashCancelAuthorisationRequest =
+  (cancelAuthorisationRequest: CancelAuthorisationRequest): string => {
+    const payloadString = JSON.stringify(cancelAuthorisationRequest);
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/authorisation.ts]: Let's use `solidityKeccak256` for serialization, as:
- it is deterministic
- it is consistent with other serializations in the application

See [example usage](https://github.com/UniversalLogin/UniversalLoginSDK/blob/2942e5c06886fd19b260bd61215b59086b3d1098/universal-login-commons/lib/core/utils/calculateSignature.ts#L11).

File: universal-login-commons/lib/core/models/authorisation.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export interface CancelAuthorisationRequest {
+  walletContractAddress: string;
+  key: string;
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/models/authorisation.ts]: Shall we add signature field here?
```ts
export interface CancelAuthorisationRequest {
  walletContractAddress: string;
  key: string;
  signature: string?;
}
```

File: universal-login-commons/lib/core/utils/authorisation.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    const payloadBytes = utils.toUtf8Bytes(payloadString);
+    return utils.keccak256(payloadBytes);
+  };
+
+export const signCancelAuthorisationRequest =
+  (cancelAuthorisationRequest: CancelAuthorisationRequest, privateKey: string): utils.Signature => {
+    const payloadDigest = hashCancelAuthorisationRequest(cancelAuthorisationRequest);
+    const signingKey = new utils.SigningKey(privateKey);
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/authorisation.ts]: 👍  Good job finding that low-level function, I was not aware it exists and created a Wallet each time.

File: universal-login-commons/lib/core/utils/authorisation.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    return signingKey.signDigest(payloadDigest);
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/authorisation.ts]: We should probably wrap the last two lines into one util `sign(payload, privateKey)` function.
Also, we should find all occurrences of Wallet.sign and replace it with a newly created `sign` function if make sense.

File: universal-login-relayer/lib/http/routes/authorisation.ts [hunk 10]
```diff
@@ -17,13 +22,30 @@ const getPending = (authorisationService : AuthorisationService) =>
...
-const denyRequest = (authorisationService : AuthorisationService) =>
-  async (data: {walletContractAddress: string, body: {key: string}}) => {
+const denyRequest = (authorisationService : AuthorisationService, provider: providers.Provider) =>
+  async (data: {walletContractAddress: string, body: {key: string, signature: utils.Signature}}) => {
+    const {walletContractAddress} = data;
+    const {key, signature} = data.body;
+
+    const cancelAuthorisationRequest: CancelAuthorisationRequest = {walletContractAddress, key};
+    const [isValid, computedAddress] = verifyCancelAuthorisationRequest(cancelAuthorisationRequest, signature, key);
+    if (!isValid) {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/routes/authorisation.ts]: You can replace it with `ensure(...)`

File: universal-login-commons/test/core/utils/authorisation.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
+import {expect} from 'chai';
+import {utils} from 'ethers';
+import {CancelAuthorisationRequest} from '../../../lib/core/models/authorisation';
+import {signCancelAuthorisationRequest, verifyCancelAuthorisationRequest, hashCancelAuthorisationRequest} from '../../../lib/core/utils/authorisation';
+
+describe('authorisation sign verify', async () => {
+  const contractAddress: string = '0x14791697260E4c9A71f18484C9f997B308e59325';
+  const privateKey = '0x9e0f0ab35e7b8d8efc554fa0e9db29235e7c52ea5e2bb53ed50d24ff7a4a6f65';
+  const address = '0xa67131F4640294a209fdFF9Ad15a409D22EEB3Dd';
+  const correctSignature: utils.Signature = {
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/authorisation.ts]: correct -> expected

File: universal-login-commons/test/core/utils/authorisation.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    r: '0x9e7a71072e325f28b36027296dae75e73b74d0e0f684047b72862229a7d4fae2',
+    s: '0x745c49528b42b6ff026d1aed5b87e612f083f2dd87ad211879fccca44caafd70',
+    v: 27
+  };
+  const correctPayloadDigest = '0x2d0177d23495e90de42405735b860ec1837aa8afcfc6ee6a45b638598db933bc';
+
+  it('Hash cancel authorisation request', async () => {
+    const cancelAuthorisationRequest: CancelAuthorisationRequest = {
+      walletContractAddress: contractAddress,
+      key: address.toLowerCase()
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/authorisation.ts]: toLowerCase -> should be done in `hashCancelAuthorisationRequest` not in test

File: universal-login-commons/lib/core/utils/authorisation.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+  };
+
+export const verifyCancelAuthorisationRequest =
+  (cancelAuthorisationRequest: CancelAuthorisationRequest, signature: utils.Signature, address: string): [boolean, string] => {
+    const payloadDigest = hashCancelAuthorisationRequest(cancelAuthorisationRequest);
+    const computedAddress = utils.recoverAddress(payloadDigest, signature);
+    return [computedAddress === address, computedAddress];
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/authorisation.ts]: You should generally never return tuples, instead use objects, eg:
```ts
return {isValid: computedAddress === address, computedAddress}
```

File: universal-login-relayer/lib/http/routes/authorisation.ts [hunk 14]
```diff
@@ -17,13 +22,30 @@ const getPending = (authorisationService : AuthorisationService) =>
...
 
-const denyRequest = (authorisationService : AuthorisationService) =>
-  async (data: {walletContractAddress: string, body: {key: string}}) => {
+const denyRequest = (authorisationService : AuthorisationService, provider: providers.Provider) =>
+  async (data: {walletContractAddress: string, body: {key: string, signature: utils.Signature}}) => {
+    const {walletContractAddress} = data;
+    const {key, signature} = data.body;
+
+    const cancelAuthorisationRequest: CancelAuthorisationRequest = {walletContractAddress, key};
+    const [isValid, computedAddress] = verifyCancelAuthorisationRequest(cancelAuthorisationRequest, signature, key);
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/routes/authorisation.ts]: `const {isValid, computedAddress} = verifyCancelAuthorisationRequest` #notuples

File: universal-login-commons/lib/core/utils/authorisation.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+export const hashCancelAuthorisationRequest =
+  (cancelAuthorisationRequest: CancelAuthorisationRequest): string => {
+    const payloadString = JSON.stringify(cancelAuthorisationRequest);
+    const payloadBytes = utils.toUtf8Bytes(payloadString);
+    return utils.keccak256(payloadBytes);
+  };
+
+export const signCancelAuthorisationRequest =
+  (cancelAuthorisationRequest: CancelAuthorisationRequest, privateKey: string): utils.Signature => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/authorisation.ts]: utils.Signature is complex type, I would serialize it to hex string (which I think is what ethers do automatically by default). #stringsignature

File: universal-login-commons/lib/core/models/authorisation.ts (hunk 7 above)
Comment (marekkirejczyk) [reply universal-login-commons/lib/core/models/authorisation.ts]: rename `key` -> `publicKey`

File: universal-login-sdk/lib/sdk.ts [hunk 16]
```diff
@@ -218,9 +219,17 @@ class UniversalLoginSDK {
     return privateKey;
   }
 
-  async denyRequest(walletContractAddress: string, key: string) {
-    await this.relayerApi.denyConnection(walletContractAddress, key);
-    return key;
+  async denyRequest(walletContractAddress: string, publicKey: string, privateKey: string) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/sdk.ts]: #stringsignature #singatureinrequest 
```ts
async denyRequest(cancelAuthorisationRequest : CancelAuthorisationRequest) ...
```

File: universal-login-sdk/lib/RelayerApi.ts [hunk 17]
```diff
@@ -39,10 +40,13 @@ export class RelayerApi {
     });
   }
 
-  async denyConnection(walletContractAddress: string, key: string) {
+  async denyConnection(walletContractAddress: string, key: string, signature: utils.Signature) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/RelayerApi.ts]: #stringsignature #singatureinrequest
```ts
async denyConnection(cancelAuthorisationRequest: CancelAuthorisationRequest) {
```

File: universal-login-relayer/lib/http/utils/sanitizers.ts [hunk 18]
```diff
@@ -24,7 +24,16 @@ export const asArrayish: Sanitizer<string | number[]> = (value, path) => {
   }
 };
 
+export const asSignature: Sanitizer<utils.Signature> = asObject({
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/utils/sanitizers.ts]: #stringsignature

File: universal-login-relayer/lib/http/routes/authorisation.ts [hunk 19]
```diff
@@ -17,13 +22,30 @@ const getPending = (authorisationService : AuthorisationService) =>
     return responseOf({ response: result });
   };
 
-const denyRequest = (authorisationService : AuthorisationService) =>
-  async (data: {walletContractAddress: string, body: {key: string}}) => {
+const denyRequest = (authorisationService : AuthorisationService, provider: providers.Provider) =>
+  async (data: {walletContractAddress: string, body: {key: string, signature: utils.Signature}}) => {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/routes/authorisation.ts]: Can we make it `data: {body: AuthorisationReuqest }`?

File: universal-login-relayer/lib/http/routes/authorisation.ts (hunk 10 above)
Comment (marekkirejczyk) [reply universal-login-relayer/lib/http/routes/authorisation.ts]: You can actually create an `ensureCancelAuthorisationRequestSignature` and push ensure to this function

File: universal-login-relayer/lib/http/routes/authorisation.ts [hunk 20]
```diff
@@ -17,13 +22,30 @@ const getPending = (authorisationService : AuthorisationService) =>
...
+    const {key, signature} = data.body;
+
+    const cancelAuthorisationRequest: CancelAuthorisationRequest = {walletContractAddress, key};
+    const [isValid, computedAddress] = verifyCancelAuthorisationRequest(cancelAuthorisationRequest, signature, key);
+    if (!isValid) {
+      throw new InvalidSignature(`cancelAuthorisationRequest failed due to invalid signature`);
+    }
+
+    const contract = new ethers.Contract(walletContractAddress, WalletMasterWithRefund.interface, provider);
+    const flatSignature = ethers.utils.joinSignature(signature);
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/routes/authorisation.ts]: This probably should be a part of new util `sign` function #stringsignature

File: universal-login-relayer/lib/http/routes/authorisation.ts [hunk 21]
```diff
@@ -17,13 +22,30 @@ const getPending = (authorisationService : AuthorisationService) =>
...
+    const payloadDigest = hashCancelAuthorisationRequest(cancelAuthorisationRequest);
+    const isCorrectAddress = await contract.isValidSignature(payloadDigest, flatSignature);
+    if (!isCorrectAddress) {
+      throw new InvalidAddress(computedAddress);
+    }
+
     const result = await authorisationService.removeRequest(data.walletContractAddress, data.body.key);
     return responseOf(result, 204);
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/routes/authorisation.ts]: Everything except the last line should probably be extracted to `AuthorisationService`.

File: universal-login-commons/lib/core/utils/authorisation.ts (hunk 8 above)
Comment (gasiortomasz) [reply universal-login-commons/lib/core/utils/authorisation.ts]: Other than that, there is lots of `const address = new Wallet(privateKey).address` lines.
Can be replaced with `const address = utils.computeAddress(privateKey)`

Comment (gasiortomasz) [reply universal-login-commons/lib/core/utils/authorisation.ts]: [example 1](https://github.com/UniversalLogin/UniversalLoginSDK/blob/1fc6e900c05cc73eaa416ddde68cad75b9007855/universal-login-commons/lib/core/utils/messages/calculateMessageSignature.ts#L41)

[example 2](https://github.com/UniversalLogin/UniversalLoginSDK/blob/1fc6e900c05cc73eaa416ddde68cad75b9007855/universal-login-wallet/src/core/services/ConnectToWallet.ts#L8)

etc..

File: universal-login-commons/lib/core/models/authorisation.ts (hunk 7 above)
Comment (gasiortomasz) [reply universal-login-commons/lib/core/models/authorisation.ts]: Agree. Also I've noticed that address and publicKey are used interchangeably 
[ekhm...](https://github.com/UniversalLogin/UniversalLoginSDK/blob/1fc6e900c05cc73eaa416ddde68cad75b9007855/universal-login-wallet/src/core/services/ConnectToWallet.ts#L8)

File: universal-login-relayer/lib/http/routes/authorisation.ts (hunk 19 above)
Comment (gasiortomasz) [reply universal-login-relayer/lib/http/routes/authorisation.ts]: Sure. What about passing provider? Kinda ugly, no?

File: universal-login-commons/lib/core/models/authorisation.ts (hunk 7 above)
Comment (marekkirejczyk) [reply universal-login-commons/lib/core/models/authorisation.ts]: By convention, we use `publicKey`, when it does not store funds.

File: universal-login-relayer/lib/core/utils/errors.ts [hunk 22]
```diff
@@ -1,15 +1,28 @@
...
 
+export class Unauthorised extends RelayerError {
+  constructor (message: string, errorType: ErrorType) {
+    super(message, errorType);
+    this.errorType = errorType;
+    Object.setPrototypeOf(this, Unauthorised.prototype);
+  }
+}
+
+export class UnathorisedAddress extends Unauthorised {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/core/utils/errors.ts]: Typo Una**u**thorised

File: universal-login-relayer/lib/core/utils/errors.ts [hunk 23]
```diff
@@ -1,15 +1,28 @@
...
+  constructor (address: string) {
+    super(`Unauthorised address: ${address}` , 'UnathorisedAddress');
+    Object.setPrototypeOf(this, UnathorisedAddress.prototype);
+  }
+}
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/core/utils/errors.ts]: Don't  forget to add support for 401 http code in `errorHandler`

File: universal-login-relayer/lib/http/routes/authorisation.ts (hunk 19 above)
Comment (marekkirejczyk) [reply universal-login-relayer/lib/http/routes/authorisation.ts]: Yup, should be `core/AuthorisationService` that handles all the work and delegated further if needed to `integration/ethereum` and `integration/sql` if needed.
Therefore route and AuthorisationService should not need provider.

File: universal-login-relayer/lib/http/middlewares/errorHandler.ts [hunk 24]
```diff
@@ -31,6 +31,8 @@ function getContentForException(err: Error) {
 export function getCodeForException(err: Error) {
   if (err instanceof ValidationFailed || err instanceof SanitizeError) {
     return 400;
+  } else if (err instanceof UnauthorisedAddress) {
+    return 401;
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/middlewares/errorHandler.ts]: Don't forget about updating test :)

File: universal-login-relayer/lib/http/routes/authorisation.ts [hunk 25]
```diff
@@ -67,7 +64,7 @@ export default (authorisationService : AuthorisationService, provider: any) => {
       walletContractAddress: asString,
       body: asObject({
         publicKey: asString,
-        signature: asSignature
+        signature: asString
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/routes/authorisation.ts]: I would keep `asSignature` - which would verify that field is a hex string of proper length.

File: universal-login-commons/lib/core/utils/authorisation.ts [hunk 26]
```diff
@@ -0,0 +1,32 @@
+import {utils} from 'ethers';
+import {CancelAuthorisationRequest} from '../models/authorisation';
+
+export const hashCancelAuthorisationRequest =
+  (cancelAuthorisationRequest: CancelAuthorisationRequest): string => {
+    const {walletContractAddress, publicKey} = cancelAuthorisationRequest;
+    return utils.solidityKeccak256(['bytes20', 'bytes20'], [walletContractAddress.toLowerCase(), publicKey.toLowerCase()]);
+  };
+
+export const sign = (payload: string, privateKey: string): string => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/authorisation.ts]: move to `commons/core/utils/sign`

File: universal-login-relayer/lib/http/routes/authorisation.ts [hunk 27]
```diff
@@ -3,10 +3,11 @@ import AuthorisationStore, {AuthorisationRequest} from '../../integration/sql/se
 import {asyncHandler, sanitize, responseOf, asString, asObject} from '@restless/restless';
 import {getDeviceInfo} from '../utils/getDeviceInfo';
 import {recoverFromCancelAuthorisationRequest, CancelAuthorisationRequest, hashCancelAuthorisationRequest} from '@universal-login/commons';
-import { ethers, providers} from 'ethers';
+import { ethers } from 'ethers';
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/http/routes/authorisation.ts]: 1) Avoid leading & trailing space
2) Why linter let it pass?

File: universal-login-commons/lib/core/utils/sign.ts [hunk 28]
```diff
@@ -0,0 +1,7 @@
+import {utils} from 'ethers';
+
+export const sign = (payload: string, privateKey: string): string => {
+    const signingKey = new utils.SigningKey(privateKey);
+    const signature = signingKey.signDigest(payload);
+    return utils.joinSignature(signature);
+  };
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/sign.ts]: unnecessary tab

File: universal-login-commons/lib/core/utils/sign.ts [hunk 29]
```diff
@@ -0,0 +1,7 @@
+import {utils} from 'ethers';
+
+export const sign = (payload: string, privateKey: string): string => {
+    const signingKey = new utils.SigningKey(privateKey);
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/sign.ts]: indentation too deep

File: universal-login-commons/lib/core/utils/sign.ts (hunk 28 above)
Comment (gasiortomasz) [reply universal-login-commons/lib/core/utils/sign.ts]: yeah, it's deleted

File: universal-login-commons/lib/core/utils/sign.ts (hunk 29 above)
Comment (gasiortomasz) [reply universal-login-commons/lib/core/utils/sign.ts]: outdated file

File: universal-login-commons/lib/core/utils/sign.ts (hunk 28 above)
Comment (gasiortomasz) [reply universal-login-commons/lib/core/utils/sign.ts]: outdated file

File: universal-login-relayer/lib/http/routes/authorisation.ts (hunk 27 above)
Comment (gasiortomasz) [reply universal-login-relayer/lib/http/routes/authorisation.ts]: I guess we should add `"check-module"` to
`"whitespace": [
      true,
      "check-branch",
      "check-decl",
      "check-operator",
      "check-separator",
      "check-type",
      "check-type-operator"
    ]`

File: universal-login-relayer/lib/integration/ethereum/services/AuthorisationService.ts [hunk 30]
```diff
@@ -0,0 +1,22 @@
+import {recoverFromCancelAuthorisationRequest, CancelAuthorisationRequest, hashCancelAuthorisationRequest} from '@universal-login/commons';
+import { ethers, providers} from 'ethers';
+import WalletMasterWithRefund from '@universal-login/contracts/build/WalletMasterWithRefund.json';
+import { UnauthorisedAddress } from '../../../core/utils/errors';
+
+class AuthorisationService {
+  constructor(private provider: providers.Provider) {}
+
+  async isValidSignature(cancelAuthorisationRequest: CancelAuthorisationRequest) {
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/integration/ethereum/services/AuthorisationService.ts]: should be ensureValidSignature (throws exception if invalid)
By convention isValidSignature return false if invalid

File: universal-login-relayer/lib/integration/ethereum/services/AuthorisationService.ts [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+    const recoveredAddress = recoverFromCancelAuthorisationRequest(cancelAuthorisationRequest);
+    const {walletContractAddress, signature} = cancelAuthorisationRequest;
+
+    const contract = new ethers.Contract(walletContractAddress, WalletMasterWithRefund.interface, this.provider);
+    const payloadDigest = hashCancelAuthorisationRequest(cancelAuthorisationRequest);
+    const isCorrectAddress = await contract.isValidSignature(payloadDigest, signature);
+    if (!isCorrectAddress) {
+        throw new UnauthorisedAddress(recoveredAddress);
+    }
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/integration/ethereum/services/AuthorisationService.ts]:
```ts
ensure(isCorrectAddress, UnauthorisedAddress, recoveredAddress)
```

File: universal-login-relayer/lib/integration/ethereum/services/AuthorisationService.ts [hunk 32]
```diff
@@ -0,0 +1,22 @@
...
+import { UnauthorisedAddress } from '../../../core/utils/errors';
+
+class AuthorisationService {
+  constructor(private provider: providers.Provider) {}
+
+  async isValidSignature(cancelAuthorisationRequest: CancelAuthorisationRequest) {
+    const recoveredAddress = recoverFromCancelAuthorisationRequest(cancelAuthorisationRequest);
+    const {walletContractAddress, signature} = cancelAuthorisationRequest;
+
+    const contract = new ethers.Contract(walletContractAddress, WalletMasterWithRefund.interface, this.provider);
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/integration/ethereum/services/AuthorisationService.ts]: Should be extracted to `integration/ethereum/WalletContract`

File: universal-login-sdk/test/authentication-relayer.ts [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+
+  beforeEach(async () => {
+    ({sdk, privateKey, contractAddress, relayer} = await loadFixture(basicSDK));
+  });
+
+  afterEach(async () => {
+    await relayer.clearDatabase();
+  });
+
+  it('Send valid cancel request', async () => {
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/authentication-relayer.ts]: Shall we make it roundtrip with creating request first? 
Here we cancel non-existing request and everything seems fine.

File: universal-login-sdk/test/authentication-relayer.ts [hunk 34]
```diff
@@ -0,0 +1,49 @@
+import {expect} from 'chai';
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/authentication-relayer.ts]: Move to test/e2e/ (along with counterfactual and sdk.ts)

File: universal-login-sdk/test/authentication-relayer.ts (hunk 33 above)
Comment (gasiortomasz) [reply universal-login-sdk/test/authentication-relayer.ts]: Hmm, the purpose of this test was to check against authorisation.

**Comment by marekkirejczyk** [comment]
Looks good, I'll merge this.

To be polished in next PR:
* Update test for errorHandler
* AuthorisationService should get AuthorisationStore as dependency and call removeRequest on it, so that denyRequest route is just one line, also integration class should be extracted for ethereum related stuff
* rename isValidSignature -> ensureValidSignature
* if () throw -> ensure
* Add a roundtrip test for requests
* Move above test to test/e2e/ (along with counterfactual and sdk.ts)

Let's discuss tomorrow

**Pull request merged**